### PR TITLE
LPS-110925 Update modal usages in fragment-web

### PR DESCRIPTION
**Pull request by markocikos**
Hey @liferay-echo ,

This is a followup on the long line of PRs. In https://github.com/wincent/liferay-portal/pull/204, we split off Babel-related tweaks, to remove it as a potential source of CI errors. It is now merged. This PR contains only update of modal for `fragment-web`, along with required API extension of modal utility. 

I tested this on a fresh `ant all`, with no issues.

History:
* [wincent#192](https://github.com/wincent/liferay-portal/pull/192)
* [jbalsas#2154](https://github.com/jbalsas/liferay-portal/pull/2154)
* [markocikos#182](https://github.com/markocikos/liferay-portal/pull/182)
* [wincent#198](https://github.com/wincent/liferay-portal/pull/198)
* [wincent#203](https://github.com/wincent/liferay-portal/pull/203)
* #156
* #218 and spin-offs:   
   * [markocikos#183](https://github.com/markocikos/liferay-portal/pull/183)
   * [markocikos#184](https://github.com/markocikos/liferay-portal/pull/184)
   * [markocikos#185](https://github.com/markocikos/liferay-portal/pull/185)
 * [wincent#204](https://github.com/wincent/liferay-portal/pull/204) (merged)

Please review the code.

Thank you!

/cc @wincent @jbalsas @john-co

### CODE REVIEW

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:stop

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:bundle

**Comment by john-co** [comment]
This seems to be related: https://github.com/liferay-echo/liferay-portal/pull/156#issuecomment-612980431 

Will investigate further: https://issues.liferay.com/browse/LRQA-57890

**Comment by john-co** [comment]
The uploadcommonfile error seems related to a change on the modal iframe element structure. It can no longer find //iframe[contains(@class,`dialog-iframe-node`)].

IframeLocator

**Comment by john-co** [comment]
May want to check the js-unit failure. That issue doesn't appear in upstream acceptance and is reproducible in https://github.com/liferay-echo/liferay-portal/pull/156

**Comment by jbalsas** [comment]
Hey @markocikos, @wincent, just a comment. I'm a bit worried that extracting the form buttons to the outside will break standard Form functionality and complicate the migration. Here are some thoughts:
- Views are unaware that they're being rendered inside an iframe and in many cases can be used in different contexts. That means the most views can't stop adding the submit button on a form without risking losing the behaviour.
- Buttons inside a form act as triggers for different form functionality. The `submit` one is obvious, but what about `reset`? What happens when a form provides a `reset` button? Is that one pushed to the footer as well? Is it triggering the right form behaviour?
- Default form functionality like "validation" is triggered differently from a form submission or a JS explicit submission. Are we properly testing those scenarios? Is the implicit form submission working (click enter inside an input in the form). Does clicking the submit button trigger the form default validation? What about our custom validation rules?

So, in summary, my main concern is:
- JS form submission !== Default form submission
- Views are unaware of where they're being rendered and might be hard to adapt to different in-iframe contexts

Just some food for thought!

**Comment by markocikos** [comment]
@john-co Should I push in your [ingore test commit](https://github.com/markocikos/liferay-portal/pull/186/commits/3b08ca0c7167b21d17855fc394698f801081cf90)? Or is this fixed in brianchandotcom#87633 as you mentioned [here](https://github.com/markocikos/liferay-portal/pull/186#issuecomment-614322397)?

**Comment by markocikos** [comment]
Hey @jbalsas ,

We will need to end up supporting both:
1. Definition of buttons through modal API.
1. Definition of buttons within iframe form.

(1) is in this PR, you can already see (2) on my [WIP branch](https://github.com/markocikos/liferay-portal/commit/3a5420402cbc3d815e160965c5aeb40b4b50eb3a).

To address individual concerns:
> * Views are unaware that they're being rendered inside an iframe and in many cases can be used in different contexts. That means the most views can't stop adding the submit button on a form without risking losing the behavior.

A valid concern. This could be worked around, but it would increase development cost. These kind of complications seemed to be more of an exception to typical use, so it shouldn't have been too bad.

> * Buttons inside a form act as triggers for different form functionality. The `submit` one is obvious, but what about `reset`? What happens when a form provides a `reset` button? Is that one pushed to the footer as well? Is it triggering the right form behaviour?

This is the big one, that made me sure that it is necessary to support (2). We have usages like multiple page tabs and complex workflows in modals. There is no way to have the same functionality without bottom-up communication from iframe to component.

> * Default form functionality like "validation" is triggered differently from a form submission or a JS explicit submission. Are we properly testing those scenarios? Is the implicit form submission working (click enter inside an input in the form). Does clicking the submit button trigger the form default validation? What about our custom validation rules?

I tested default validation and it is working correctly on a typical form. I would not put my hand in fire that there no edge case there where it fails. But, from what I've seen, typically when we have complex submit-related behavior, the submit event is intercepted, followed by complex preparation logic and `submitForm`. This would still work.

I did not test implicit form submission, it is not applicable to this PR. I suspect it does not work, but it can easily be patched in the util solution.



So, if we are planning to support (2), why do we still want to support (1)? It is because I feel we want to promote (1) as a solution:
- Placement of buttons in a modal footer better matches their concern. They structurally _should_ be there. If they are in iframe form, they are a part of the modal body.
- When buttons are placed in a modal footer, Clay determines their styles. Clay follows Lexicon guidelines correctly, while iframe buttons are slightly off. It is a duplication of code.
- While I understand we need to support stuff like tabs and workflows in a modal, we should not promote it. Ideally, following [Lexicon guidelines](https://liferay.design/lexicon/core-components/modals/#usage), these cases should be rewritten as pages. Modals should be simple.

In addition, we would need to support (1) simply because AUI modal supported it, and we have 15 usages of that API in portal. 

/cc @wincent

**Comment by john-co** [comment]
@markocikos, it’s fixed.

**Comment by markocikos** [comment]
I rebased and force-pushed, no additional changes.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by ealonso** [comment]
@markocikos @john-co @jbalsas @wincent
please see: https://github.com/liferay-echo/liferay-portal/pull/255

**Comment by markocikos** [comment]
@ealonso Looks good! We can close this PR, right?

**Comment by ealonso** [comment]
@markocikos I think so
